contributed_definitions/nyaml/NXmpes.yaml

```diff
@@ -289,34 +289,92 @@ NXmpes(NXobject):
       (NXmanipulator):
         exists: optional
         doc: |
-          Manipulator for positioning of the sample.
-        sample_temperature(NX_FLOAT):
+          Manipulator for positioning of the sample.      
+        temperature_sensor(NXsensor):
           exists: recommended
-          unit: NX_TEMPERATURE
-        drain_current(NX_FLOAT):
-          exists: recommended
-          unit: NX_CURRENT
-        sample_bias(NX_FLOAT):
+          name:
+            exists: recommended
+          measurement:
+            enumeration: [temperature]
+          type:
+            exists: optional
+          value(NX_FLOAT):
+            exists: recommended
+        sample_heater(NXactuator):
           exists: recommended
-          unit: NX_CURRENT
-          doc:
-          - |
-            Bias of the sample with respect to analyser ground.
-            
-            This field may also be found in NXsample if present.
-          - |
-            xref:
-              spec: ISO 18115-1:2023
-              term: 8.41
-              url: https://www.iso.org/obp/ui/en/#iso:std:iso:18115:-1:ed-3:v1:en:term:8.41
-        device_information(NXfabrication):
+          name:
+            exists: recommended
+          acutation:
+            enumeration: [temperature]
+          type:
+            exists: optional
+          setpoint(NX_FLOAT):
+            exists: recommended
```

Comment:
In this current setup, is there actually a place to e.g. store the actual heater power as function of time? This reads like a temperature setpoint, which is not the same.

Reply:
You are right, this is currently missing.

I think having a setpoint(_log) here that stores the setpoint values as a function of time is reasonable (as it is). There should then be an additional field (and its log) that stores the heater power, or similar.  I think this is why there was an additional `output` field in `NXacutator` at the beginning. However, `output` is probably not the best term. Any good suggestions? Maybe generic `PARAMETER` and `PARAMETER_log`? And then specify whenever `NXactuator `is used?

Reply:
Strictly speaking, the setpoint of the sample heater as a "Temperature" requires the use of some controller that reads in the temperature and sets the heater power. So The actuator should write "heater_power", and the setpoint should be in NXpid. But not sure if that makes sense to break things up so much...

Reply:
How about this:
**NXacutator.yaml**
```yaml
NXactuator(NXobject):
  OUTPUT(NX_FLOAT):
    unit: NX_ANY
    doc: |
      Any output that the actuator produces.
      For example, a heater can have the field heater_power(NX_FLOAT).
  OUTPUT_log(NXlog):
    unit: NX_ANY
    doc: |
      Time history of actuator outputs.
  (NXpid):
    doc: |
      If the actuator is PID-controlled, the settings of the PID controller can be stored here.
    setpoint(NX_FLOAT):
      unit: NX_ANY
      doc: |
        Nominal actuator setpoint.
        Can be a scalar or a vector (of [n] actuations).
    setpoint_log(NXlog):
      doc: |
        Time history of actuator setpoints.
```

**NXpid.yaml**
```yaml
NXpid(NXobject):
  setpoint(NX_FLOAT):
    unit: NX_ANY
    doc: |
      The Setpoint(s) used as an input for the PID controller.
      
      It can also be a link to an NXsensor.value field.
  [...]
  setpoint_log(NXlog):
    unit: NX_ANY
    doc: |
      Time log of the setpoint(s) used as an input for the PID controller.
```
And then to be used in **NXmanipulator.yamö**:
```yaml
NXmanipulator(NXobject):
  [...]
  sample_heater(NXactuator):
    doc: Device to heat the sample.
    actuation:
      enumeration: [temperature]
    heater_power(NX_FLOAT):
      unit: NX_ANY
      doc: |
        In case of a fixed or averaged heating power, this is the scalar heater power.
        It can also be a 1D array of heater powers (without time stamps).
    heater_power_log(NXlog):
      unit: NX_ANY
       doc: |
         In the case of an experiment in which the heater power is changed and recorded with time stamps,
         this is an array of length m of temperature setpoints.
    (NXpid):
      setpoint(NX_FLOAT):
        unit: NX_TEMPERATURE
        doc: |
          In case of a fixed or averaged temperature, this is the scalar temperature setpoint.
          It can also be a 1D array of temperature setpoints (without time stamps).
      setpoint_log(NXlog):
        value(NX_NUMBER):
          unit: NX_TEMPERATURE
          doc: |
            In the case of an experiment in which the temperature is changed and the setpoints are
            recorded with time stamps, this is an array of length m of temperature setpoints.
```
@domna also curious about your thoughts.

Reply:
> How about this: **NXacutator.yaml**
> 
> ```yaml
> NXactuator(NXobject):
>   OUTPUT(NX_FLOAT):
>     unit: NX_ANY
>     doc: |
>       Any output that the actuator produces.
>       For example, a heater can have the field heater_power(NX_FLOAT).
>   OUTPUT_log(NXlog):
>     unit: NX_ANY
>     doc: |
>       Time history of actuator outputs.
> ...
> ```

This design actually means that every field of type `NX_FLOAT` in the NXactuator class is an output, because there is now way of distinguishing this. If this is fine this is ok, but maybe it's better to have some kind of prefix? i.e., `output_OUTPUT` and `output_OUTPUT_log`?

I also thought whether `actuation` is a good vocabulary for the quantity being actuated? Maybe something like `physical_quantity` is a bit easier to understand from it's name, since it's already clear that we are dealing with something which is actuated in this group.

Other than those two points I like the general design.

Reply:
> This design actually means that every field of type NX_FLOAT in the NXactuator class is an output, because there is now way of distinguishing this. If this is fine this is ok, but maybe it's better to have some kind of prefix? i.e., output_OUTPUT and output_OUTPUT_log?

I already thought about this. `output_OUTPUT_log` seems very abstract and not easy to understand. There are currently no other field of type `NX_FLOAT` in NXactuator, so my proposal would still work. I guess if the base class is extended later, this could be problematic.

> I also thought whether actuation is a good vocabulary for the quantity being actuated? Maybe something like physical_quantity is a bit easier to understand from it's name, since it's already clear that we are dealing with something which is actuated in this group.

I chose `actuation` to mirror `measurement` in `NXsensor`. But I agree, `physical_quantity` probably makes more sense. I have renamed it.

Reply:
> > This design actually means that every field of type NX_FLOAT in the NXactuator class is an output, because there is now way of distinguishing this. If this is fine this is ok, but maybe it's better to have some kind of prefix? i.e., output_OUTPUT and output_OUTPUT_log?
> 
> I already thought about this. `output_OUTPUT_log` seems very abstract and not easy to understand. There are currently no other field of type `NX_FLOAT` in NXactuator, so my proposal would still work. I guess if the base class is extended later, this could be problematic.

Actually, this would not be problematic for the definitions, as new fields can just be added. However, if someone wants to store a field of type `NX_FLOAT` in NXactuator which is not in this base class it would automatically be assumed that this is an `OUTPUT` field. I think this could be very confusing for nexus users.
> 
> > I also thought whether actuation is a good vocabulary for the quantity being actuated? Maybe something like physical_quantity is a bit easier to understand from it's name, since it's already clear that we are dealing with something which is actuated in this group.
> 
> I chose `actuation` to mirror `measurement` in `NXsensor`. But I agree, `physical_quantity` probably makes more sense. I have renamed it.

👍️

Reply:
Should we not also change measurement in NXsensor then to physical_quantity?

Reply:
Otherwise, I find this solution good, and don't see a problem with the potentially assignement of undocumented fields. It's the same in NXdata or NXtransformations, no?

Reply:
> Otherwise, I find this solution good, and don't see a problem with the potentially assignement of undocumented fields. It's the same in NXdata or NXtransformations, no?

That's true it's handled the same there. Then let's keep it like this.

Reply:
> Should we not also change measurement in NXsensor then to physical_quantity?

This breaks the NIAC standard, so I didn't do it.

Reply:
Ah, right, this is not our's...
Well then, let's keep it like it is.